kvm/msr.go

```diff
@@ -164,3 +168,118 @@ func GetMSRFeatureIndexList(kvmFd uintptr, list *MSRList) error {
 
 	return err
 }
+
+type MSREntry struct {
+	Index   uint32
+	padding uint32
+	Data    uint64
+}
+
+type MSRS struct {
+	NMSRs   uint32
+	padding uint32
+	Entry   MSREntry
```

Comment:
Maybe it would be better to use slice here since it is variable length.
https://www.kernel.org/doc/html/v6.3/virt/kvm/api.html#kvm-get-msrs

Reply:
That's what I meant with my first comment. 
I refere this code:
https://github.com/qemu/qemu/blob/d145c0da22cde391d8c6672d33146ce306e8bf75/target/i386/kvm/kvm.c#L493

It is the only instance it is used in QEMU, and it works as I implemented it here. The moment I implement it with dynamic sizing it does not work at all. (I suspect the documentation is "We wish it would work like this" and QEMU shows how it actually work.

Reply:
Additionally I found:

https://github.com/qemu/qemu/blob/d145c0da22cde391d8c6672d33146ce306e8bf75/target/i386/kvm/kvm.c#L3052

Reply:
Hmmm, I tried a little.

I read the QEMU code. I think QEMU is making requests one entry at a time for some reason.
But the following code, for example, can support variable-length entries.


<details>
  <summary>Click to expand</summary>
  
```diff
ubuntu2204:~/gokvm$ git --no-pager diff a4202f4844edba40230f00c61748f103ab70dd05
diff --git a/kvm/kvm_test.go b/kvm/kvm_test.go
index be681fc447d1..50ecabf22a91 100644
--- a/kvm/kvm_test.go
+++ b/kvm/kvm_test.go
@@ -1428,15 +1428,22 @@ func TestGetSetMSRS(t *testing.T) {

        msrs := &kvm.MSRS{
                NMSRs: list.NMSRs,
-               Entry: kvm.MSREntry{
-                       Index: list.Indicies[0],
-               },
+       }
+       msrs.Entries = make([]kvm.MSREntry, list.NMSRs)
+       for i := 0; i< int(list.NMSRs); i++ {
+               msrs.Entries[i] = kvm.MSREntry{
+                        Index: list.Indicies[i],
+               }
        }

        if err := kvm.GetMSRs(vcpuFd, msrs); err != nil {
                t.Fatal(err)
        }

+       for _, entry := range msrs.Entries {
+               t.Logf("%#v\n", entry)
+       }
+
        if err := kvm.SetMSRs(vcpuFd, msrs); err != nil {
                t.Fatal(err)
        }
diff --git a/kvm/msr.go b/kvm/msr.go
index d4ca5c5eb63b..8fbc3ed05730 100644
--- a/kvm/msr.go
+++ b/kvm/msr.go
@@ -167,14 +167,14 @@ func GetMSRFeatureIndexList(kvmFd uintptr, list *MSRList) error {

 type MSREntry struct {
        Index   uint32
-       padding uint32
+       Padding uint32
        Data    uint64
 }

 type MSRS struct {
        NMSRs   uint32
-       padding uint32
-       Entry   MSREntry
+       Padding uint32
+       Entries []MSREntry
 }

 func (m *MSRS) Bytes() ([]byte, error) {
@@ -184,20 +184,22 @@ func (m *MSRS) Bytes() ([]byte, error) {
                return nil, err
        }

-       if err := binary.Write(&buf, binary.LittleEndian, m.padding); err != nil {
+       if err := binary.Write(&buf, binary.LittleEndian, m.Padding); err != nil {
                return nil, err
        }

-       if err := binary.Write(&buf, binary.LittleEndian, m.Entry.Index); err != nil {
-               return nil, err
-       }
+       for _, entry := range m.Entries {
+               if err := binary.Write(&buf, binary.LittleEndian, entry.Index); err != nil {
+                       return nil, err
+               }

-       if err := binary.Write(&buf, binary.LittleEndian, m.Entry.padding); err != nil {
-               return nil, err
-       }
+               if err := binary.Write(&buf, binary.LittleEndian, entry.Padding); err != nil {
+                       return nil, err
+               }

-       if err := binary.Write(&buf, binary.LittleEndian, m.Entry.Data); err != nil {
-               return nil, err
+               if err := binary.Write(&buf, binary.LittleEndian, entry.Data); err != nil {
+                       return nil, err
+               }
        }

        return buf.Bytes(), nil
@@ -215,19 +217,12 @@ func NewMSRS(data []byte) (*MSRS, error) {
                return nil, err
        }

-       if err := binary.Read(&buf, binary.LittleEndian, &m.padding); err != nil {
+       if err := binary.Read(&buf, binary.LittleEndian, &m.Padding); err != nil {
                return nil, err
        }

-       if err := binary.Read(&buf, binary.LittleEndian, &m.Entry.Index); err != nil {
-               return nil, err
-       }
-
-       if err := binary.Read(&buf, binary.LittleEndian, &m.Entry.padding); err != nil {
-               return nil, err
-       }
-
-       if err := binary.Read(&buf, binary.LittleEndian, &m.Entry.Data); err != nil {
+       m.Entries = make([]MSREntry, m.NMSRs)
+       if err := binary.Read(&buf, binary.LittleEndian, &m.Entries); err != nil && !errors.Is(err, io.EOF) {
                return nil, err
        }

@@ -243,7 +238,7 @@ func SetMSRs(vcpuFd uintptr, msrs *MSRS) error {
        }

        if _, err := Ioctl(vcpuFd,
-               IIOW(kvmSetMSRS, unsafe.Sizeof(len(data))),
+               IIOW(kvmSetMSRS, 8),
                uintptr(unsafe.Pointer(&data[0]))); err != nil {
                return err
        }
@@ -266,7 +261,7 @@ func GetMSRs(vcpuFd uintptr, msrs *MSRS) error {
        }

        if _, err := Ioctl(vcpuFd,
-               IIOWR(kvmGetMSRS, unsafe.Sizeof(msrs)),
+               IIOWR(kvmGetMSRS, 8),
                uintptr(unsafe.Pointer(&data[0]))); err != nil {
                return err
        }

```
</details> 

From the output below, gokvm is able to get the MSR data (Data 0x7040600070406 for MSR index 0x277) from the KVM API.

```bash
ubuntu2204:~/gokvm$ sudo go test . /kvm/...  -run TestGetSetMSRS -v
...
    kvm_test.go:1444: kvm.MSREntry{Index:0x277, Padding:0x0, Data:0x7040600070406}
```

Reply:
Ok, so if this works as well I will implement it. I tried several things myself and coundnt make it work. Nice you managed to do so :) I happily apply the changes.